homeassistant/components/reolink/binary_sensor.py

```diff
@@ -0,0 +1,172 @@
+"""This component provides support for Reolink binary sensors."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from reolink_aio.api import (
+    FACE_DETECTION_TYPE,
+    PERSON_DETECTION_TYPE,
+    PET_DETECTION_TYPE,
+    VEHICLE_DETECTION_TYPE,
+    Host,
+)
+
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+    BinarySensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ReolinkData
+from .const import DOMAIN
+from .entity import ReolinkCoordinatorEntity
+
+
+@dataclass
+class ReolinkBinarySensorEntityDescriptionMixin:
+    """Mixin values for Reolink binary sensor entities."""
+
+    value: Callable[[Host, int | None], bool]
+
+
+@dataclass
+class ReolinkBinarySensorEntityDescription(
+    BinarySensorEntityDescription, ReolinkBinarySensorEntityDescriptionMixin
+):
+    """A class that describes binary sensor entities."""
+
+    icon: str = "mdi:motion-sensor"
+    icon_off: str = "mdi:motion-sensor-off"
+    supported: Callable[[Host, int | None], bool] = lambda host, ch: True
+
+
+BINARY_SENSORS = (
+    ReolinkBinarySensorEntityDescription(
+        key="motion",
+        name="Motion",
+        device_class=BinarySensorDeviceClass.MOTION,
+        value=lambda api, ch: api.motion_detected(ch),
+    ),
+    ReolinkBinarySensorEntityDescription(
+        key=FACE_DETECTION_TYPE,
+        name="Face",
+        icon="mdi:face-recognition",
+        device_class=BinarySensorDeviceClass.MOTION,
```

Comment:
Is this still a motion sensor? Are we saying it's a face detected sensor that is triggered by a motion event?

Reply:
All these sensors are provided by the reolink API, they all come from the camera directly.
Reolink analyses the video stream of the camera internally inside the camera and uses AI algoritims to figure out what it in the image: face, people, dog/cat, vehicle. The motion is also derived from the camera image by the camera internally (I think by checking the change in the overall picture with some algoritm taking into account brightness or something).

The only exception is the "visitor" sensor, this is the physical button on a video doorbell. So this binary sensor shows when a doorbell is beeing pressed down. (you get API webhook calls for visitor: true and visitor: false.

Reply:
@MartinHjelmare schould I remove the device_class?

Reply:
When does the feature corresponding to these binary sensors reset? Is it a continuous processing and the camera creates states depending on the result of the processing? Ie, as long as there's a face in the images, the camera says `face state: true`, and when there's no face anymore in the images, the camera says `face state: false`?

Reply:
Yes it is a continuous processing, as long as there is a face/person/dog_cat/vehicle in the image the value will remain True in the API when polled.
You get one ONVIF event notification when the value changes from False to True and you get a second notification when the value changes from True to False.
2x same notification, but the "value" parameter in the XML will be "true" or "false" (string).

Reply:
Exemple of the ONVIF event notification that we receive through a webhook for when a person is detected for the first time in the image:

```
<?xml version="1.0" encoding="UTF-8"?>
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope" xmlns:SOAP-ENC="http://www.w3.org/2003/05/soap-encoding" xmlns:c14n="http://www.w3.org/2001/10/xml-exc-c14n#" xmlns:chan="http://schemas.microsoft.com/ws/2005/02/duplex" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:ns1="http://www.onvif.org/ver20/media/wsdl" xmlns:saml1="urn:oasis:names:tc:SAML:1.0:assertion" xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:tad="http://www.onvif.org/ver20/analytics/wsdl" xmlns:tdn="http://www.onvif.org/ver10/network/wsdl" xmlns:tds="http://www.onvif.org/ver10/device/wsdl" xmlns:ter="http://www.onvif.org/ver10/error" xmlns:tev="http://www.onvif.org/ver10/events/wsdl" xmlns:timg="http://www.onvif.org/ver20/imaging/wsdl" xmlns:tmd="http://www.onvif.org/ver10/deviceIO/wsdl" xmlns:tns1="http://www.onvif.org/ver10/topics" xmlns:tptz="http://www.onvif.org/ver20/ptz/wsdl" xmlns:trc="http://www.onvif.org/ver10/recording/wsdl" xmlns:trp="http://www.onvif.org/ver10/replay/wsdl" xmlns:trt="http://www.onvif.org/ver10/media/wsdl" xmlns:trv="http://www.onvif.org/ver10/receiver/wsdl" xmlns:tse="http://www.onvif.org/ver10/search/wsdl" xmlns:tt="http://www.onvif.org/ver10/schema" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsa5="http://www.w3.org/2005/08/addressing" xmlns:wsc="http://docs.oasis-open.org/ws-sx/ws-secureconversation/200512" xmlns:wsdd="http://schemas.xmlsoap.org/ws/2005/04/discovery" xmlns:wsnt="http://docs.oasis-open.org/wsn/b-2" xmlns:wsrfbf="http://docs.oasis-open.org/wsrf/bf-2" xmlns:wsrfr="http://docs.oasis-open.org/wsrf/r-2" xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wstop="http://docs.oasis-open.org/wsn/t-1" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:xenc="http://www.w3.org/2001/04/xmlenc#" xmlns:xop="http://www.w3.org/2004/08/xop/include" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
   <SOAP-ENV:Header>
      <wsa5:Action SOAP-ENV:mustUnderstand="1">http://docs.oasis-open.org/wsn/bw-2/NotificationConsumer/Notify</wsa5:Action>
   </SOAP-ENV:Header>
   <SOAP-ENV:Body>
      <wsnt:Notify>
         <wsnt:NotificationMessage>
            <wsnt:Topic Dialect="http://www.onvif.org/ver10/tev/topicExpression/ConcreteSet">tns1:RuleEngine/MyRuleDetector/PeopleDetect</wsnt:Topic>
            <wsnt:Message>
               <tt:Message UtcTime="2023-01-18T11:20:55Z" PropertyOperation="Changed">
                  <tt:Source>
                     <tt:SimpleItem Name="Source" Value="000" />
                  </tt:Source>
                  <tt:Data>
                     <tt:SimpleItem Name="State" Value="true" />
                  </tt:Data>
               </tt:Message>
            </wsnt:Message>
         </wsnt:NotificationMessage>
      </wsnt:Notify>
   </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

Reply:
Ok. Then I think it's ok to make binary sensor entities to represent these API features.

I'd only mark the motion feature as motion device class.

Reply:
Thank you!
I have removed the device class from the AI binary sensors.